File: Data Structures & Algorithms/permutations/submission-1.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        int n = nums.size();
        vector<vector<int>> ans;
        vector<int> curr;
        solve(0,curr,ans, nums);
        return ans;
    }

    void solve(int idx,vector<int>& curr, vector<vector<int>>& ans, vector<int>& nums){
        int n = nums.size();
        if (idx==n){
            ans.push_back(nums);
            return;
        }
        for (int i=idx; i<n; i++){
            swap(nums[idx],nums[i]);
            solve(idx+1,curr,ans,nums);
            swap(nums[idx],nums[i]);
        }
    }


};

File: Data Structures & Algorithms/subsets/submission-1.go
func subsets(nums []int) [][]int {
	var ans [][]int
	var curr []int
	generateSubsets(0, &curr, nums, &ans)
	return ans
}

func generateSubsets(idx int, curr *[]int, nums []int, ans *[][]int) {
	if idx == len(nums) {
		tmp := make([]int, len(*curr))
		copy(tmp, *curr)
		*ans = append(*ans, tmp)
		return
	}
	*curr = append(*curr, nums[idx])
	generateSubsets(idx+1, curr, nums, ans)
	*curr = (*curr)[:len(*curr)-1]
	generateSubsets(idx+1, curr, nums, ans)
}
